### PR TITLE
feat(margin): add `margin` property to lvgl

### PR DESCRIPTION
**Pull request by W-Mai**
### Description of the feature or fix

After a fierce battle of ideas, as a former front-end developer, I still wanted to add the `margin` attribute to `lvgl`, but of course adding the `margin` attribute has no effect on those who don't use it.

In the end I introduced the `margin` property to `lvgl`. now the following method has been added

```cpp
void lv_obj_set_style_margin_top(struct _lv_obj_t * obj, lv_coord_t value, lv_style_selector_t selector);
void lv_obj_set_style_margin_bottom(struct _lv_obj_t * obj, lv_coord_t value, lv_style_selector_t selector);
void lv_obj_set_style_margin_left(struct _lv_obj_t * obj, lv_coord_t value, lv_style_selector_t selector);
void lv_obj_set_style_margin_right(struct _lv_obj_t * obj, lv_coord_t value, lv_style_selector_t selector);
void lv_obj_set_style_margin_ver(struct _lv_obj_t * obj, lv_coord_t value, lv_style_selector_t selector);
void lv_obj_set_style_margin_hor(struct _lv_obj_t * obj, lv_coord_t value, lv_style_selector_t selector);
void lv_obj_set_style_margin_all(struct _lv_obj_t * obj, lv_coord_t value, lv_style_selector_t selector);

void lv_obj_get_style_margin_top(struct _lv_obj_t * obj, lv_coord_t value, lv_style_selector_t selector);
void lv_obj_get_style_margin_bottom(struct _lv_obj_t * obj, lv_coord_t value, lv_style_selector_t selector);
void lv_obj_get_style_margin_left(struct _lv_obj_t * obj, lv_coord_t value, lv_style_selector_t selector);
void lv_obj_get_style_margin_right(struct _lv_obj_t * obj, lv_coord_t value, lv_style_selector_t selector);
```

And the `lv_obj_align` and `lv_obj_align_to` methods have been enhanced to support the `margin` property.
Similarly, `lv_obj_scroll` has been adjusted to increase their support for `margin`.
Corresponding changes have also been made to the `LV_SIZE_CONTENT` behavior of the component.

Glad if you can accept the margin attribute back into lvgl.

Here is a screenshot of the demo:
without margin:
image
with margin:
image

I tested several demos of `lvgl` and found no impact on the existing behavior.

### Checkpoints
- [x] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [x] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [ ] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [ ] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [ ] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [x] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [x] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [x] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [x] Use typed pointers instead of `void *` pointers
- [x] Do not `malloc` into a static or global variables. Instead declare the variable in `LV_ITERATE_ROOTS` list in [`lv_gc.h`](https://github.com/lvgl/lvgl/blob/master/src/misc/lv_gc.h) and mark the variable with `GC_ROOT(variable)` when it's used.
- [x] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [x] Widget members function must start with `lv_<modul_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [x] `struct`s should be used via an API and not modified directly via their elements.
- [x] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [x] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [x] Arguments must be named in H files too.
- [x] To register and use callbacks one of the followings needs to be followed: 
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

### CODE REVIEW

**Comment by W-Mai** [comment]
https://github.com/lvgl/lvgl/issues/3763

**Comment by W-Mai** [comment]
If it's acceptable, I'll introduce margin into lv_flex next.

**Comment by kisvegabor** [comment]
I was thinking a lot of whether adding margin or not.  I've also collected some references (other libs, articles, etc) where there is no margin but in the end only one thing matters: do we need it to build UIs? 

So I'm ok with adding it if you can show a few **real-life** UI designs (i.e. not just boxes as demonstration)  where margin is really useful. My experience is that (while creating embedded UI all day) flex and grid (even in their current form) are enough to manage layouts.

**Comment by W-Mai** [comment]
It's hard for me to find a situation that meets your requirements, because the vast majority of websites use margin.
You can see the existence of `margin` on any website.

So I wrote a simple demo to demonstrate the advantages of `margin` in some cases.
The animation is as follows.

https://user-images.githubusercontent.com/24841247/202697380-5a060159-1257-4450-9ed4-85dd4f8e552d.mov

If you use `gap` to implement it, you can't control each item finely and individually, and if you use `transition`, it won't affect the layout of other item.

And offset is not a `style` property, it cannot be used in flex layouts.
At the same time, this example can also easily modify the distribution of item.

The code is also very simple, using `attribute animation` can be achieved.

```cpp
static lv_style_transition_dsc_t dsc;
static const lv_style_prop_t props[] = {
    LV_STYLE_MARGIN_TOP,
    LV_STYLE_MARGIN_BOTTOM,
    LV_STYLE_MARGIN_LEFT,
    LV_STYLE_MARGIN_RIGHT,
    LV_STYLE_HEIGHT,
    LV_STYLE_PROP_INV,
};
static lv_style_t style;
lv_style_init(&style);

lv_style_transition_dsc_init(&dsc, props, lv_anim_path_ease_in_out, 50, 0, nullptr);
lv_style_set_transition(&style, &dsc);
lv_style_set_margin_top(&style, 10);
lv_style_set_margin_left(&style, 10);
lv_style_set_margin_right(&style, 10);
lv_style_set_margin_bottom(&style, 10);
lv_style_set_height(&style, 20);

......

lv_obj_add_style(o, &style, LV_STATE_PRESSED);
```

In addition, `margin` has an advantage over `offset` because `margin` is a property that takes effect after it is **set once**.

As a front-end developer, seeing `margin` is more cordial than other attributes and can adapt to `lvgl` development more quickly.

The design drafts provided by UI designers are also based on `margin`, `padding`, `border` and so on.

To web, iOS 's `UIKit` and `SwiftUI` also have `margin` attributes, which is very easy to use.

You can say that there are other ways to replace `margin` to achieve a similar effect. What I want to say is that if you have `margin`, you don't have to **work around** and you can manipulate each obj in a more straightforward way.

Adding margin **will not** put more burden on `lvgl`, on the contrary, it is a more convenient tool for more developers.
\\(^o^)/~

**Comment by kisvegabor** [comment]
> So I wrote a simple demo to demonstrate the advantages of margin in some cases.

Actually it's a good example.  However a little bit special. Without margin I would solve it by wrapping each colored boxes into a transparent box and change the size of the outer transparent box.

But there is one thing that I don't understand in this demo. At 0:04 you click the pink box it stays in place. But when you blue box that stays in place.  But the length of the track became larger, with the extra margin and when this larger width is centered all elements should have moved a little. It's like the pivot point of the centering is the click position.   What is trick, or am I missing something obvious? :slightly_smiling_face: 

> To web, iOS 's UIKit and SwiftUI also have margin attributes, which is very easy to use.

There are other frameworks where there is no margin. See for example Framer, a very new website designer tool: https://www.youtube.com/watch?v=2JAyFZOXxDw

> The design drafts provided by UI designers are also based on margin, padding, border and so on.

I have a quite negative feeling about the current state of the web development. I think it's quite messy and hacky. So I would afraid of taking it as a  good reference (rather the opposite, we should learn the mistakes of web). I consider it only as a source of ideas which should be taken by a huge grain of salt.

> Adding margin will not put more burden on lvgl, on the contrary, it is a more convenient tool for more developers.

Addig margin actually has consequences even if one doesn't want to use them:
- steeper learning curve on layouts, it can be mixed with padding, longer docs
- more style property to `get` might cause  performance drop if the layouts with a lot of widgets need to be arrange.  
- makes flex and grid algorithm more complex (more possible bugs, harder to maintain, corner cases)
- mixing padding and margin can result in hard to debug layouts 
- margin collapse?

My opinion is that `margin` is a conceptually bad idea because it acts out of the element. E.g. you create a style for a button with `margin-top:10px`. It means literally whatever is on top of the button I'd like to keep 10px distance. How do you know that it will true in general? Keep distance from whatever? Doesn't sound like a good design philosophy. Can't we use "spacers" in these rare cases instead (like Framer does)?

On the other hand padding is enclosed into the object. You know what you place inside and you can adjust padding  accordingly.

In short, I think `margin` is a bad solution to the "how to hack the spacing around a given element" problem. Grid should handle these cases very well when there is a fixed number of elements and you would like to craft a well-designed unique layout (e.g. a card with avatar + text + button). So IMO the use of grid can replace margin very well.  

Anyway I'm really open to discuss the need of margin I just had a little bit less popular opinion about it. :slightly_smiling_face:  I suggest approaching this from "what is the concrete problem we would like to solve" and find a solution to the problem (I.e. this design is hard to implement now, how to make it simpler?), instead of adding features which might be redundant or not really needed.

**Comment by xiaoxiang781216** [comment]
@kisvegabor since margin is part of CSS spec, many designer and developer get used to develop UI layout with margin property. Actually, we are developing a JavaScript UI framework base on LVGL for smart watch. It's very important to make the transition as smooth as possible for the traditional web developer, so we decide to follow CSS spec as close as possible. Because LVGL layout lack the support of margin(and other) property, we have to disable all LVGL layout engine and layout(base on https://github.com/facebook/yoga) the widget out side LVGL. But this approach has some bad side effect in the production environment, so the better approach is to improve lv_flex to support the key CSS property.
Margin is one key issue which block we use lv_flex directly, so it's important to see margin included in LVGL layout system, @kisvegabor . 🙂

**Comment by kisvegabor** [comment]
I understand your situation and I don't want to argue about your decision of following CSS. Probably it's a complex system and there are no simple answers. :slightly_smiling_face: 
I also don't want to repeat myself about IMO how bad CSS is. 

So let's approach it constructively! What is the problem and what how can we solve it?

**Problem**
LVGL is CSS-like so it sounds reasonable to assume there is `margin` too. It breaks portability between web-like designs and embedded. So the question is not really "how to implement UIs without margin" but "have margin because CSS has it too". 

**Options**
1. Support `margin` everywhere just like CSS. I'm definitely against it because it brings a lot of complication everywhere. E.g. a simple `lv_obj_set_pos()` shouldn't care about `margin` 
2. Support margin only in `flex` and `grid`. It makes more sense as it's a local complication.
4. Offer a straight forward `margin` replacement workaround.  Usually I wrap the widget into a transparent container and set the `padding` on the wrapper. It has the same effect as `margin`. It works with LVGL only but it messes up the hierarchy and it's hacky.

**Solution**
I could be ok with 2) if we really keep is as local as possible:
- only for flex and grid
- no margin collapse
- maybe make it optional? On flex now we need to get only `flex_grow` and margin means 4 extra property to get for each object. On layouts with 1000 items it might matter.
- shall we apply `margin` on `%` values too? E.g. with `width=100%` it might make sense to use margin.

**Comment by W-Mai** [comment]
https://github.com/lvgl/lvgl/pull/3845

File: scripts/style_api_gen.py [hunk 1]
```diff
@@ -99,6 +99,23 @@
  'style_type': 'num',   'var_type': 'lv_coord_t', 'default':0, 'inherited': 0, 'layout': 1, 'ext_draw': 0,
  'dsc': "Sets the padding between the columns. Used by the layouts."},
 
+{'section': 'Margin', 'dsc' : "Properties to describe spacing around an object. Very similar to the margin properties in HTML."},
+{'name': 'MARGIN_TOP',
+ 'style_type': 'num',   'var_type': 'lv_coord_t',  'default':0, 'inherited': 0, 'layout': 1, 'ext_draw': 0,
+ 'dsc': "Sets the margin on the top. It makes the object area bigger in this direction."},
```
Comment (kisvegabor) [inline scripts/style_api_gen.py]:
```suggestion
 'dsc': "Sets the margin on the top. The object will keep this space from its siblings in layouts. "},
```

**Comment by kisvegabor** [comment]
I've checked the changes and they look good in general. Thank you! 

A few things:
- With percentage size margin is still no considered. See [here](https://github.com/lvgl/lvgl/blob/master/src/core/lv_obj_pos.c#L99)
- Do you plan to add margin to grid too?
- It'd be really nice to have some unit tests for simple sizing/positioning, flex and grid. Are you interested in adding them? I think it'd be helpful during the development too and I assume you already have some test codes.

**Comment by W-Mai** [comment]
> * With percentage size margin is still no considered. See [here](https://github.com/lvgl/lvgl/blob/master/src/core/lv_obj_pos.c#L99)

Aiming at the problem of `percentage` size, since the current padding of lvgl does not support percentage, I want to open a separate pr to support both `padding` and `margin`, but percentage may need some processing when it comes to problems such as `LV_SIZE_CONT` that need to calculate the size of the parent component.

> * Do you plan to add margin to grid too?

I have this plan, but I haven't looked at the code for the grid layout yet. I'll add margin to the grid layout later.

> * It'd be really nice to have some unit tests for simple sizing/positioning, flex and grid. Are you interested in adding them? I think it'd be helpful during the development too and I assume you already have some test codes.

I'm interested in it. I can try it.

**Comment by kisvegabor** [comment]
> Aiming at the problem of percentage size, since the current padding of lvgl does not support percentage, I want to open a separate pr to support both padding and margin, but percentage may need some processing when it comes to problems such as LV_SIZE_CONT that need to calculate the size of the parent component.

I meant`margin=20px` on a child with `width = 100%` should make the child smaller. (Not `margin = 10%`, but it would be great too.)

**Comment by kisvegabor** [comment]
I think only a few unit tests needs to be added and we are ready to go.

**Comment by W-Mai** [comment]
I wrote a simple testcase and wondered if it would work

**Comment by kisvegabor** [comment]
I don't know why but it seems the the new test file is not picked by CI. See https://github.com/lvgl/lvgl/actions/runs/3891743797/jobs/6642347842

Maybe `test_grid()`, `test_margin()` and `test_align()` should be in there own file.

**Comment by W-Mai** [comment]
> I don't know why but it seems the the new test file is not picked by CI. See https://github.com/lvgl/lvgl/actions/runs/3891743797/jobs/6642347842
> 
> Maybe `test_grid()`, `test_margin()` and `test_align()` should be in there own file.

Does this result mean being picked by ci?
[398](https://github.com/lvgl/lvgl/actions/runs/3891743797/jobs/6642347842#step:5:399)
> [ 82%] Generating test_margin_Runner.c
[ 82%] Building C object CMakeFiles/test_line.dir/src/test_cases/test_line.c.o
[ 82%] Building C object CMakeFiles/test_margin.dir/src/test_cases/test_margin.c.o
[ 82%] Building C object CMakeFiles/test_line.dir/test_line_Runner.c.o
[ 82%] Building C object CMakeFiles/test_margin.dir/test_margin_Runner.c.o
[ 83%] Linking C executable test_line
[ 84%] Linking C executable test_margin
[ 84%] Built target test_line
[ 84%] Generating test_mem_Runner.c
[ 84%] Built target test_margin
[ 85%] Generating test_msg_Runner.c

> 18/37 Test #19: test_line ........................   Passed    0.07 sec
19/37 Test #20: test_margin ......................   Passed    0.40 sec

**Comment by kisvegabor** [comment]
Ah, I see 2 problem:
1. Seemingly I'm blind
2. Searching with Ctrl+F in the CI log is not working well as only parts of the log is loaded. 

:smile: 

Regardless to the CI picked it up correctly, just for the sake of cleaner code, could you still separate the test into 3 files?

**Comment by W-Mai** [comment]
> Ah, I see 2 problem:
> Seemingly I'm blind
> Searching with Ctrl+F in the CI log is not working well as only parts of the log is loaded.

hah, I had the same problem before😁

**Comment by kisvegabor** [comment]
Just one last suggestion about the docs.  If you agree with it please update the docs and after I can merge this PR.

File: scripts/style_api_gen.py (hunk 1 above)
Comment (W-Mai) [reply scripts/style_api_gen.py]: Haha, I'm a Chinese with poll English.😁
That's great.

I should update all 4 description, right?

**Comment by W-Mai** [comment]
> Just one last suggestion about the docs. If you agree with it please update the docs and after I can merge this PR.

Well done.

**Comment by kisvegabor** [comment]
Great, thank you much! :tada: :tada: :tada: 
I was really against margin but if use it wisely it can really help :relaxed:
